subworkflows/local/call_variants.nf

```diff
@@ -78,21 +79,17 @@ workflow CALL_VARIANTS {
             .set { ch_varcallerinfo }
 
         ADD_VARCALLER_TO_BED(ch_varcallerinfo)
-        ch_bed_in_annot = ADD_VARCALLER_TO_BED.out.gz.collect()
-        ch_tbi_in_annot = ADD_VARCALLER_TO_BED.out.tbi.collect()
-
+        bed_with_tbi = ADD_VARCALLER_TO_BED.out.gz_tbi.map{ meta, gz, tbi -> return [ gz,tbi ] }
         in_bcftools_annot = REMOVE_DUPLICATES.out.vcf
             .join(REMOVE_DUPLICATES.out.tbi)
-            .join(ch_bed_in_annot)
-            .join(ch_tbi_in_annot)
+            .combine(bed_with_tbi)
```

Comment:
Will this work as intended? Looking at the docs for combine: https://www.nextflow.io/docs/latest/reference/operator.html#combine

Seems like it will produce all combinations, i.e. if you have

```
vcf_channel = [[f1.vcf, f1.vcf.tbi], [f2.vcf, f2.vcf.tbi]]
bed_channel = [[f1.bed, f1.bed.tbi], [f2.bed, f2.bed.tbi]]
```

Then combining these:

```
vcf_channel.combine(bed_channel)
```

Would yield:

```
[
[ [f1.vcf, f1.vcf.tbi], [f1.bed, f1.bed.tbi]],
[ [f1.vcf, f1.vcf.tbi], [f2.bed, f2.bed.tbi]],
[ [[f2.vcf, f2.vcf.tbi], [f1.bed, f1.bed.tbi]],
[ [f2.vcf, f2.vcf.tbi], [f2.bed, f2.bed.tbi]],
]
```

While I suspect you want only:

```
[
[ [f1.vcf, f1.vcf.tbi], [f1.bed, f1.bed.tbi]],
[ [[f2.vcf, f2.vcf.tbi], [f1.bed, f1.bed.tbi]],
]
```

Maybe I am missing something. Just started pondering seeing this, and wanted to ask!

Reply:
Perhaps this block of code could be replaced by something simpler, to avoid the joins and combines all together. All it does is to append a `FOUND_IN=variant_caller` to the INFO field and add a header line, right? https://github.com/genomic-medicine-sweden/tomte/blob/dd179c76e7f13176f13cf3ebd76993b5c3f3f465/subworkflows/local/call_variants.nf#L76-L90 


In nallo I made [one module](https://github.com/genomic-medicine-sweden/nallo/blob/dev/modules/local/add_found_in_tag/main.nf) that takes a VCF and variant_caller value as input, but it could perhaps also be done in a small python script, or with [vcfexpress](https://github.com/brentp/vcfexpress). Then you could just pass the output from the caller through this process, e.g. like [this](https://github.com/genomic-medicine-sweden/nallo/blob/42ec8147ca1f333d09888b34a1cc72a0a4fc84a2/subworkflows/local/call_repeat_expansions/main.nf#L44-L47) in nallo:

```
    ADD_FOUND_IN_TAG (
        TRGT.out.vcf.map { meta, vcf -> [ meta, vcf, [] ] },
        "TRGT"
    )
```

What do you think @Lucpen?

Reply:
I completely see your point, I am looking more into it but when I dump the channels, this is how they look:

```
vcf_channel =[[case:finequagga, id:ACC5963A2, sample:ACC5963A2, strandedness:reverse, sex:NA, single_end:false, fq_pairs:1], /path/ACC5963A2_split_rmdup.vcf.gz, /path/ACC5963A2_split_rmdup.vcf.gz.tbi] 

bed_channel=[/path/bcftools.bed.gz, /path/bcftools.bed.gz.tbi]

vcf_channel.combine(bed_channel)

[[case:finequagga, id:ACC5963A2, sample:ACC5963A2, strandedness:reverse, sex:NA, single_end:false, fq_pairs:1], /path/ACC5963A2_split_rmdup.vcf.gz, /hpath/ACC5963A2_split_rmdup.vcf.gz.tbi, /path/bcftools.bed.gz, /path/bcftools.bed.gz.tbi]

```

Reply:
> I completely see your point, I am looking more into it but when I dump the channels, this is how they look:
> 
> ```
> vcf_channel =[[case:finequagga, id:ACC5963A2, sample:ACC5963A2, strandedness:reverse, sex:NA, single_end:false, fq_pairs:1], /path/ACC5963A2_split_rmdup.vcf.gz, /path/ACC5963A2_split_rmdup.vcf.gz.tbi] 
> 
> bed_channel=[/path/bcftools.bed.gz, /path/bcftools.bed.gz.tbi]
> 
> vcf_channel.combine(bed_channel)
> 
> [[case:finequagga, id:ACC5963A2, sample:ACC5963A2, strandedness:reverse, sex:NA, single_end:false, fq_pairs:1], /path/ACC5963A2_split_rmdup.vcf.gz, /hpath/ACC5963A2_split_rmdup.vcf.gz.tbi, /path/bcftools.bed.gz, /path/bcftools.bed.gz.tbi]
> ```

Looking at the VCF channel, it looks like you just have one VCF and one BED at the moment. Here I think it would give the desired number of combinations, i.e. 1 x 1 = 1. 

I think `.combine` will start giving you additional combinations when you have 2+ sets of data in each channel 🤔 You could try running with two samples and dump.

Reply:
I.e. running this in `nextflow console`:

```
vcf_channel = Channel.of(["f1.vcf", "f1.vcf.tbi"], ["f2.vcf", "f2.vcf.tbi"])
bed_channel = Channel.of(["f1.bed", "f1.bed.tbi"], ["f2.bed", "f2.bed.tbi"])
combined = vcf_channel.combine(bed_channel)
combined.view()
```

Gives:

```
[f1.vcf, f1.vcf.tbi, f1.bed, f1.bed.tbi]
[f2.vcf, f2.vcf.tbi, f1.bed, f1.bed.tbi]
[f1.vcf, f1.vcf.tbi, f2.bed, f2.bed.tbi]
[f2.vcf, f2.vcf.tbi, f2.bed, f2.bed.tbi]
```

Reply:
I think its a great suggestion, I will try to implement it :)

Reply:
[Done](https://github.com/genomic-medicine-sweden/tomte/pull/184)!